File: 6.zigzag-conversion/solution.go
func convert(s string, numRows int) string {
	if numRows == 1 {
		return s
	}

	var rows []string

	for i := 0; i < numRows && i < len(s); i++ {
		rows = append(rows, "")
	}

	curRow := 0
	goingDown := false

	for _, c := range s {
		rows[curRow] += string(c)
		if curRow == 0 || curRow == numRows-1 {
			goingDown = !goingDown
		}
		if goingDown {
			curRow++
		} else {
			curRow--
		}
	}

	ret := ""
	for _, row := range rows {
		ret += row
	}
	return ret
}

File: 6.zigzag-conversion/solution.ts
function convert(s: string, numRows: number): string {
    if (numRows == 1) {
        return s
    }

    const rows: string[] = [];

    for (let i = 0; i < numRows && i < s.length; i++) {
        rows.push("");
    }

    let curRow = 0;
    let goingDown = false;

    for (const c of s) {
        rows[curRow] += c;
        if (curRow == 0 || curRow == numRows-1) {
            goingDown = !goingDown;
        }
        if (goingDown) curRow++;
        else curRow--;
    }

    return rows.join("");
};